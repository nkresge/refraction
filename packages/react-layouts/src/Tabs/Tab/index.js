/**
 * Copyright (c) Quid, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
// @flow
import * as React from 'react';
import css from '@emotion/css/macro';
import styled from '@emotion/styled/macro';
import { textStyles, withFallback as wf, sizes } from '@quid/theme';

export type Props = {
  children: React.Node,
  id: string,
  isActive: boolean,
  onClick: Function,
};

const TabButton = styled.button`
  ${textStyles('normal', 'bold')};
  border: 0;
  font-size: inherit;
  font-family: inherit;
  cursor: pointer;
  background-color: inherit;

  position: relative;
  flex: 1;
  padding: ${sizes.regular};
  line-height: 1;

  &:focus:not(:focus-visible) {
    outline: 0;
  }
  &:focus-visible {
    z-index: 1;
  }

  ${wf(props => {
    return (
      props.isActive &&
      css`
        background-color: ${props.theme.primaryInverse};
        color: ${props.theme.selected};
      `
    );
  })};
`;

const Tab = styled(({ isActive, id, onClick, children, ...props }: Props) => {
  return (
    <li {...props}>
      <TabButton
        isActive={isActive}
        id={`${id}_tab`}
        aria-controls={`${id}_panel`}
        aria-selected={isActive}
        role="tab"
        onClick={onClick}
      >
        {children}
      </TabButton>
    </li>
  );
})`
  flex: 1;
  display: flex;
`;

export default Tab;
