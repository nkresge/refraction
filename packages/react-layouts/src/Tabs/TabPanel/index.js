/**
 * Copyright (c) Quid, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
// @flow
import React from 'react';
import styled from '@emotion/styled/macro';
import { sizes } from '@quid/theme';

export type Props = {
  label: string,
  id: string,
  className?: string,
};

const TabPanel = styled(({ label, id, className, ...props }: Props) => (
  <div id={`panel_${id}`} aria-labelledby={`tab_${id}`} {...props} />
))`
  padding: calc(${sizes.regular} + 3px);
`;

export default TabPanel;
