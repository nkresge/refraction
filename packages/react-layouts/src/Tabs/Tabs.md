`Tabs` organize content across different screens, data sets, and other interactions.

```js
initialState = {
  activeTab: 'two',
};

<Tabs
  id="tab_a"
  activeTab={state.activeTab}
  onChange={activeTab => setState({ activeTab })}
  sideTabsContent={
    <Button>
      <Icon name="expand" />
    </Button>
  }
>
  <Tabs.TabPanel label="One" name="one">
    Tabs content example 1. <br /> <br /> Lorem ipsum dolor sit amet,
    consectetur adipiscing elit. Proin vestibulum ipsum ut rhoncus tincidunt.
    Etiam lacinia nunc sit amet imperdiet tempus. Donec dapibus magna quis enim
    tincidunt, vel ultrices massa viverra. Vestibulum ante ipsum primis in
    faucibus orci luctus et ultrices posuere cubilia Curae; Integer ac arcu quis
    risus mollis placerat. Suspendisse nec fermentum lacus. Ut eleifend tempus
    tellus, ut malesuada odio sagittis in. Cras id pretium velit. Suspendisse
    fringilla leo a massa efficitur, vel aliquet metus gravida.
  </Tabs.TabPanel>
  <Tabs.TabPanel label="Two" name="two">
    Tabs content example 2. <br /> <br /> Vestibulum sollicitudin ligula nec
    lectus sagittis, a luctus nisl rutrum. Duis vitae eros nunc. Aliquam sed
    varius arcu. Vestibulum scelerisque rutrum accumsan. Proin nec metus non
    massa venenatis viverra eu eget odio. Maecenas eget vehicula velit.
    Pellentesque ornare non libero id venenatis. Nunc tortor neque, bibendum
    vitae turpis vel, porta faucibus ligula. Integer risus nulla, ornare ut
    lorem sed, semper gravida dui. Fusce a arcu at erat consequat pretium quis a
    mauris.
  </Tabs.TabPanel>
  <Tabs.TabPanel label="Three" name="three">
    Tabs content example 3. <br /> <br /> Vivamus pretium metus orci, et
    tristique est congue eu. Sed condimentum enim pharetra sagittis vestibulum.
    Maecenas in quam nec metus ultricies porttitor at in purus. Aenean at
    porttitor sapien. Nunc ac egestas odio. Class aptent taciti sociosqu ad
    litora torquent per conubia nostra, per inceptos himenaeos. Aenean tincidunt
    facilisis venenatis. Aliquam erat volutpat. Nam pretium et erat eget
    dapibus. Morbi facilisis augue eu nisi venenatis, egestas ultricies justo
    fringilla. Donec facilisis ipsum urna, nec tincidunt ante consequat non.
  </Tabs.TabPanel>
</Tabs>;
```
