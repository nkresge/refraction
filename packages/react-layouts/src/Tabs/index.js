/**
 * Copyright (c) Quid, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
// @flow
import * as React from 'react';
import styled from '@emotion/styled/macro';
import css from '@emotion/css/macro';
import {
  Tabs as RTPTabs,
  TabList as RTPTabList,
  TabPanel as RTPTabPanel,
} from '@quid/react-tabs-provider';
import TabPanel from './TabPanel';
import Tab from './Tab';

export type Props = {
  children?: React.Node,
  id: string,
  onChange: Function,
  activeTab?: string,
  sideTabsContent?: React.Node,
  className?: string,
};

const TabList = styled.ul`
  display: flex;
  flex-shrink: 0;
  padding: 0;
  ${props =>
    css`
      border-bottom: 1px solid ${props.theme.primaryInverse};
    `};
`;

const TabPanels = styled.div`
  display: flex;
  position: relative;
  transition: height 0.2s ease-in-out;
  flex: 1;
  overflow: auto;
`;

// There's something wrong with Enzyme + jest-emotion that prevents us to use
// React.Fragment as child of the Popper component in this instance
// This is a dirty workaround
// istanbul ignore next
const DevFragment =
  process.env.NODE_ENV === 'test' ? 'x-fragment' : React.Fragment;

const Tabs = styled(
  ({
    id,
    children,
    activeTab,
    sideTabsContent,
    onChange,
    className,
    ...props
  }: Props) => {
    return (
      <div {...props}>
        <RTPTabs active={activeTab} onSelect={e => onChange(e)}>
          {({ active }) => (
            <DevFragment>
              <RTPTabList>
                {({ select, active }) => (
                  <TabList role="tablist">
                    {React.Children.map(children, (child, index) => {
                      const isActive = child.props.name === activeTab;
                      return (
                        <Tab
                          key={index}
                          id={`tab_${id}_${index}`}
                          isActive={isActive}
                          onClick={() => select(child.props.name)}
                        >
                          {child.props.label}
                        </Tab>
                      );
                    })}
                    {sideTabsContent}
                  </TabList>
                )}
              </RTPTabList>
              <TabPanels>
                {React.Children.map(children, (child, index) => {
                  return (
                    <RTPTabPanel name={child.props.name}>
                      {({ name }) => child}
                    </RTPTabPanel>
                  );
                })}
              </TabPanels>
            </DevFragment>
          )}
        </RTPTabs>
      </div>
    );
  }
)`
  display: flex;
  flex-direction: column;
`;

Tabs.TabPanel = TabPanel;
Tabs.Tab = Tab;

export default Tabs;
