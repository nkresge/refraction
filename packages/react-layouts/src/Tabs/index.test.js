/**
 * Copyright (c) Quid, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
// @flow
import React from 'react';
import { mount } from 'enzyme';
import Tabs from './index';
const { Tab, TabPanel } = Tabs;

it('should match TabPanel', () => {
  const wrapper = mount(
    <TabPanel label="foo" name="0" id="0">
      foo
    </TabPanel>
  );

  expect(wrapper).toMatchSnapshot();
});

it('should match a Tabs example', () => {
  const wrapper = mount(
    <Tabs activeTab="one" id="id">
      <TabPanel label="One" name="one">
        foo
      </TabPanel>
      <TabPanel label="Two" name="two">
        bar
      </TabPanel>
    </Tabs>
  );

  expect(wrapper).toMatchSnapshot();
});

it('should call onChange on tab click', () => {
  const handleChange = jest.fn();
  const wrapper = mount(
    <Tabs activeTab="0" id="id" onChange={handleChange}>
      <TabPanel label="foo" name="0" id="0">
        foo
      </TabPanel>
      <TabPanel label="bar" name="1" id="1">
        bar
      </TabPanel>
    </Tabs>
  );
  wrapper
    .find(Tab)
    .at(1)
    .find('button')
    .simulate('click');
  expect(handleChange).toBeCalledWith('1');
});
