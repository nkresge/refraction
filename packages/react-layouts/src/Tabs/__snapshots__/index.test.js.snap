// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match TabPanel 1`] = `
<TabPanel
  id="0"
  label="foo"
  name="0"
>
  <div
    aria-labelledby="tab_0"
    id="panel_0"
    name="0"
  >
    foo
  </div>
</TabPanel>
`;

exports[`should match a Tabs example 1`] = `
.emotion-8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  padding: 0;
  border-bottom: 1px solid;
}

.emotion-2 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-0 {
  font-family: IBM Plex Sans,Lucida Grande,Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 1.57;
  color: #2E3338;
  font-size: 14px;
  line-height: 1.57;
  color: #2E3338;
  font-weight: bold;
  border: 0;
  font-size: inherit;
  font-family: inherit;
  cursor: pointer;
  background-color: inherit;
  position: relative;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  padding: 10px;
  line-height: 1;
  background-color: #FFFFFF;
  color: #00c1bb;
}

.emotion-0:focus:not(:focus-visible) {
  outline: 0;
}

.emotion-0:focus-visible {
  z-index: 1;
}

.emotion-4 {
  font-family: IBM Plex Sans,Lucida Grande,Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 1.57;
  color: #2E3338;
  font-size: 14px;
  line-height: 1.57;
  color: #2E3338;
  font-weight: bold;
  border: 0;
  font-size: inherit;
  font-family: inherit;
  cursor: pointer;
  background-color: inherit;
  position: relative;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  padding: 10px;
  line-height: 1;
}

.emotion-4:focus:not(:focus-visible) {
  outline: 0;
}

.emotion-4:focus-visible {
  z-index: 1;
}

.emotion-10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  -webkit-transition: height 0.2s ease-in-out;
  transition: height 0.2s ease-in-out;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  overflow: auto;
}

<Tabs
  activeTab="one"
  id="id"
>
  <div>
    <x-fragment>
      <TabList
        role="tablist"
      >
        <ul
          className="emotion-8 emotion-9"
          role="tablist"
        >
          <Tab
            id="tab_id_0"
            isActive={true}
            key="0/.0"
            onClick={[Function]}
          >
            <li
              className="emotion-2 emotion-3"
            >
              <TabButton
                aria-controls="tab_id_0_panel"
                aria-selected={true}
                id="tab_id_0_tab"
                isActive={true}
                onClick={[Function]}
                role="tab"
              >
                <button
                  aria-controls="tab_id_0_panel"
                  aria-selected={true}
                  className="emotion-0 emotion-1"
                  id="tab_id_0_tab"
                  onClick={[Function]}
                  role="tab"
                >
                  One
                </button>
              </TabButton>
            </li>
          </Tab>
          <Tab
            id="tab_id_1"
            isActive={false}
            key="1/.1"
            onClick={[Function]}
          >
            <li
              className="emotion-2 emotion-3"
            >
              <TabButton
                aria-controls="tab_id_1_panel"
                aria-selected={false}
                id="tab_id_1_tab"
                isActive={false}
                onClick={[Function]}
                role="tab"
              >
                <button
                  aria-controls="tab_id_1_panel"
                  aria-selected={false}
                  className="emotion-4 emotion-1"
                  id="tab_id_1_tab"
                  onClick={[Function]}
                  role="tab"
                >
                  Two
                </button>
              </TabButton>
            </li>
          </Tab>
        </ul>
      </TabList>
      <TabPanels>
        <div
          className="emotion-10 emotion-11"
        >
          <TabPanel
            label="One"
            name="one"
          >
            <div
              aria-labelledby="tab_undefined"
              id="panel_undefined"
              name="one"
            >
              foo
            </div>
          </TabPanel>
        </div>
      </TabPanels>
    </x-fragment>
  </div>
</Tabs>
`;
